src/KubernetesClient/KubernetesClientConfiguration.ConfigFile.cs

```diff
@@ -557,19 +558,32 @@ public static ExecCredentialResponse ExecuteExternalCommand(ExternalExecution co
                 throw new KubeConfigException($"external exec failed due to: {ex.Message}");
             }
 
-            var stdout = process.StandardOutput.ReadToEnd();
-            var stderr = process.StandardError.ReadToEnd();
-            if (string.IsNullOrWhiteSpace(stderr) == false)
+
+            var stdOutTask = process.StandardOutput.ReadToEndAsync(); // Assumes process exits
+
+            var buffer = new char[4096];
+            var stdErrorTask = process.StandardError.ReadAsync(buffer, 0, buffer.Length); // Assumes process will continue
+
+            var result = Task.WaitAny(new Task[] { stdErrorTask, stdOutTask }, TimeSpan.FromSeconds(10));
+
+            if (result == -1)
             {
-                throw new KubeConfigException($"external exec failed due to: {stderr}");
+                throw new KubeConfigException("external exec failed due to timeout");
             }
+            else if (result == 0)
+            {
+                var stderrBuilder = new StringBuilder();
+                stderrBuilder.Append(buffer, 0, stdErrorTask.Result);
 
-            // Wait for a maximum of 5 seconds, if a response takes longer probably something went wrong...
-            process.WaitForExit(5);
+                if (stderrBuilder.Length > 0)
+                {
+                    throw new KubeConfigException($"external exec failed due to: {stderrBuilder}");
+                }
+            }
 
             try
             {
-                var responseObject = KubernetesJson.Deserialize<ExecCredentialResponse>(stdout);
+                var responseObject = KubernetesJson.Deserialize<ExecCredentialResponse>(stdOutTask.Result);
```

Comment:
`.Result` vs `getawait()`

see https://stackoverflow.com/questions/17284517/is-task-result-the-same-as-getawaiter-getresult

for better explanation